efficient_attention/SONIC/sonic_lin_eva_mha.py

```diff
@@ -0,0 +1,309 @@
+                                               nn.LayerNorm(self.head_dim)])
+        else:
+            raise NotImplementedError
+
+        if self.use_t5_rpe:
+            self.rel_pos_bias = T5RelativePositionBias(self.scale,
+                                                       num_heads=self.num_heads,
+                                                       causal=False,
+                                                       num_buckets=max(min(int((self.window_size + self.ext_size) / 2), 64), 16),
+                                                       max_distance=self.window_size + self.ext_size)
+
+        self.random_rng = jax.random.PRNGKey(0)
+
+    def window_partition(self, x, shape, ext_window_size, pad_val=0, window_size=None):
+        if window_size is None:
+            window_size = self.window_size
+        if self.attn_2d:
+            raise NotImplementedError
+        else:
+            return window_1d_partition(
+                x,
+                window_size=window_size,
+                ext_window_size=ext_window_size,
+                pad_val=pad_val
+                )
+
+    def window_merge(self, x, shape, window_size=None):
+        if window_size is None:
+            window_size = self.window_size
+        if self.attn_2d:
+            '''
+            assert isinstance(shape, (list, tuple))
+            output = window_2d_merge(
+                x,
+                window_size=window_size, 
+                hw_tuple=shape
+                )
+            return rearrange(output, '... H W d ->... (H W) d')
+            '''
+            raise NotImplementedError
+        else:
+            return window_1d_merge(x)
+
+    def _process_input(self, x, key_padding_mask=None):
+        B, N, C = x.shape
+        if self.attn_2d:
+            raise NotImplementedError
+        else:
+            if self.window_size > 0:
+                if key_padding_mask is None:
+                    x, key_padding_mask = pad_to_multiple(x, self.window_size, dim=-2, create_mask=True)
+                else:
+                    x = pad_to_multiple(x, self.window_size, dim=-2)
+                    key_padding_mask = pad_to_multiple(key_padding_mask, self.window_size, dim=-1, value=True)
+                N = x.shape[-2]
+        return x, key_padding_mask, N
+
+
+    def __call__(self, x, switch: bool, *, train, key_padding_mask=None):
+        ## Jax complains about passing in multiple arguments.
+        ## So we do the hack of concatenating the queries, keys and values into a list and unpacking it.
+        mask_val = -5e4
+        query, key, value = x
+        batch_size, sequence_length, hidden_dims = query.shape
+        # pdb.set_trace()
+
+        assert all(len(i.shape) == 3 for i in x), "Incorrect size of input, should be [batch, seq length, hidden dimension]"
+        if key.shape[1] == value.shape[1] == 128 and self.up_train and switch:
```

Comment:
This if condition is slightly incorrect. 

Look at the if statement in the other SONIC variants and follow suite. Basically if `self.up_train` is `false` we want to default to the Lin-Whatever combination.